gh-115999: Specialize `LOAD_GLOBAL` in free-threaded builds

mpage opened the pull request:
Enable specialization of `LOAD_GLOBAL` in free-threaded builds.  

Thread-safety of specialization in free-threaded builds is provided by the following:

- A critical section is held on both the globals and builtins objects during specialization. This ensures we get an atomic view of both builtins and globals during specialization.
- Generation of new keys versions is made atomic in free-threaded builds.
- Existing helpers are used to atomically modify the opcode.

Thread-safety of specialized instructions in free-threaded builds is provided by the following:

- Relaxed atomics are used when loading and storing dict keys versions. This avoids potential data races as the dict keys versions are read without holding the dictionary's per-object lock in version guards.
- Dicts keys objects are passed from keys version guards to the downstream uops. This ensures that we are loading from the correct offset in the keys object. Once a unicode key has been stored in a keys object for a combined dictionary in free-threaded builds, the offset that it is stored in will never be reused for a different key. Once the version guard passes, we know that we are reading from the correct offset.
- The dictionary read fast-path is used to read values from the dictionary once we know the correct offset.

Performance-wise, this looks like a [3-4% improvement](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241108-3.14.0a1+-0e50451-NOGIL) on free-threaded builds on pyperformance.
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

mpage committed:
Make specialization of LOAD_GLOBAL thread-safe

Thread-safety of specialization in free-threaded builds:

- A critical section is held on both the globals and builtins objects during
  specialization. This ensures we get an atomic view of both builtins and globals
  during specialization.
- Generation of new keys versions is made atomic in free-threaded builds.
- We use helpers safely modify the bytecode.

Thread-safety of specialized instructions in free-threaded builds:

- Dicts keys objects are passed from keys version guards to the downstream uops.
  This ensures that we are loading from the correct offset in the keys object.
  Once a unicode key has been stored in a keys object for a combined dictionary
  in free-threaded builds, the offset that it is stored in will never be reaused
  for a different key.
- The dictionary read fast-path is used to read values from the dictionary.
- Relaxed atomics are used when loading and storing dict keys versions. This
  avoids potential data races as the dict keys versions may be read without
  holding the dictionary's per-object lock while the instructions are executing.

mpage committed:
_Py_GetExecutor needs a CS in free-threaded builds

This handles the case where another thread is instrumenting the bytecode.

mpage committed:
Undo workaround for cases generator bug

mpage committed:
Fix unused function warning

mpage committed:
Double check that keys are still valid

mpage committed:
Fix formatting

mpage committed:
Fix linkage

mpage committed:
`_Py_TryIncrefCompare` can't escape

Fidget-Spinner reviewed (commented):
This was less complicated than I initially anticipated. Nice!

LGTM overall, but I'll withold from approval to let others review it first.

colesbury commented on Objects/dictobject.c:
```diff
@@ -4498,7 +4498,7 @@ dict_popitem_impl(PyDictObject *self)
         }
     }
-    self->ma_keys->dk_version = 0;
+    FT_ATOMIC_STORE_UINT32_RELAXED(self->ma_keys->dk_version, 0);
```
There's one more assignment to `dk_version` in `insert_split_key` that should use this pattern.

colesbury commented on Python/specialize.c:
```diff
@@ -1520,104 +1520,103 @@ PyObject *descr, DescriptorClassification kind, bool is_method)
-void
-_Py_Specialize_LoadGlobal(
+static void
+specialize_load_global_lock_held(
```
We should mark the dicts as "shared" (`_PyObject_GC_SET_SHARED` / `SET_DICT_SHARED`) when we specialize for load global. That ensures that the keys object will not be immediately freed when the dictionary is resized, which is important if another thread is trying to read the keys without holding a lock.

The dict code enforces this with `ensure_shared_on_read()`, but I think we should have the version check subsume this: if the dict has a valid version, then it's marked as shared (and we can attempt the non-locking read without checking the owning thread).

So maybe ensure the dict is shared in `_PyDictKeys_GetVersionForCurrentState`.

colesbury commented on Python/bytecodes.c:
```diff
@@ -1599,47 +1599,70 @@ dummy_func(
+            if (globals_keys != _Py_atomic_load_ptr_acquire(&dict->ma_keys)) {
+                Py_DECREF(res_o);
+                DEOPT_IF(true);
+            }
```
I *think* these checks that the keys haven't changed are not necessary. We have them in `dictobject.c` too, and I think they're not necessary there either.

I think the checks were inherited from the nogil forks, which used a different scheme for dictionary keys.

(Also the use of `global_keys` here seems to contradict the `DEAD(globals_keys)` above.)

colesbury commented on Python/bytecodes.c:
```diff
@@ -1599,47 +1599,70 @@ dummy_func(
             SYNC_SP();
             DEOPT_IF(res_o == NULL);
+            #if Py_GIL_DISABLED
+            int increfed = _Py_TryIncrefCompare(&entries[index].me_value, res_o);
```
I think we want to support deferred reference counting here. Loading functions to call them will use `LOAD_GLOBAL_MODULE` and `LOAD_GLOBAL_BUILTINS`.

I don't think we have the equivalent of `_Py_TryIncrefCompare` for `_PyStackRef` yet, it needs to be written.

colesbury reviewed (commented):
Nice! A few comments below.

mpage committed:
Use atomics for setting keys version for split keys

mpage committed:
Ensure dict is shared when assigning keys version

mpage committed:
Don't check keys after loading them

Keys are freed using QSBR. Nothing occurs in between loading the keys
and loading the value that would cause the QSBR version to advance,
ensuring that the keys object cannot be freed and reused.

mpage committed:
Support deferred refcounting

mpage committed:
Merge branch 'main' into gh-115999-tlbc-load-global

colesbury commented on Objects/dictobject.c:
```diff
@@ -1285,6 +1285,20 @@ ensure_shared_on_resize(PyDictObject *mp)
+{
+    ASSERT_DICT_LOCKED((PyObject *) mp);
+    #ifdef Py_GIL_DISABLED
+    if (!_Py_IsOwnedByCurrentThread((PyObject *)mp) && !IS_DICT_SHARED(mp)) {
```
I think this check should just be: `if (!IS_DICT_SHARED(mp))`. 

The `_Py_IsOwnedByCurrentThread()` check here might be okay with some other changes and a few assumptions (like specialization remains per-thread), but I think it's easier to reason about if this is just marks the dict as shared if it's not already.

mpage replied:
🤦 yep

colesbury commented on Objects/dictobject.c:
```diff
@@ -7313,20 +7327,67 @@ _PyDictKeys_DecRef(PyDictKeysObject *keys)
+{
+    ASSERT_DICT_LOCKED((PyObject *) dict);
+    assign_keys_version_result res = assign_keys_version(interp, dict->ma_keys);
+    if (res.assigned) {
```
I don't think you need the `res.assigned` check and the combination with `_Py_IsOwnedByCurrentThread` in `ensure_shared_on_keys_version_assignment` is not correct, I think: we could end up not marking the dict as shared in a non-owning thread.

Marking the dict as shared is cheap and this is not called frequently, and I expect the code will be simpler without the `assign_keys_version_result`.

colesbury commented on Include/internal/pycore_dict.h:
```diff
@@ -90,6 +90,15 @@ extern PyObject *_PyDict_FromKeys(PyObject *, PyObject *, PyObject *);
+ *
+ * Returns the version number, or zero if it was not possible to get a version number. */
+extern uint32_t _PyDict_GetKeysVersionForCurrentState(
+        PyInterpreterState *interp, PyDictObject *dictkeys);
```
s/dictkeys/dict

mpage committed:
Fix param name

mpage committed:
Always mark keys shared

mpage committed:
Update comment

mpage committed:
Merge branch 'main' into gh-115999-tlbc-load-global

mpage committed:
Don't pass reason to unspecialize

mpage committed:
Merge branch 'main' into gh-115999-tlbc-load-global
